tests/aws/services/apigateway/test_apigateway_lambda_cfn.py

```diff
@@ -1,80 +1,304 @@
+import base64
+import json
+import time
+import uuid
+
 import aws_cdk as cdk
 import aws_cdk.aws_apigateway as apigateway
+import aws_cdk.aws_iam as iam
 import aws_cdk.aws_lambda as awslambda
 import pytest
 
 from localstack.testing.pytest import markers
+from localstack.utils.strings import short_uid
+from localstack.utils.sync import retry
+from localstack.utils.testutil import get_lambda_log_events
+
+API_DESTINATION_AUTHS = [
+    {
+        "type": "BASIC",
+        "key": "BasicAuthParameters",
+        "parameters": {"Username": "user", "Password": "pass"},
+    },
+    {
+        "type": "API_KEY",
+        "key": "ApiKeyAuthParameters",
+        "parameters": {"ApiKeyName": "Api", "ApiKeyValue": "apikey_secret"},
+    },
+    {
+        "type": "OAUTH_CLIENT_CREDENTIALS",
+        "key": "OAuthParameters",
+        "parameters": {
+            # AuthorizationEndpoint will be replaced dynamically below
+            "AuthorizationEndpoint": "replace_this",
+            "ClientParameters": {"ClientID": "id", "ClientSecret": "password"},
+            "HttpMethod": "POST",
+            "OAuthHttpParameters": {
+                "BodyParameters": [
+                    {"Key": "grant_type", "Value": "client_credentials"},
+                    {"Key": "client_id", "Value": "id"},
+                    {"Key": "client_secret", "Value": "password"},
+                    {"Key": "oauthbody", "Value": "value1"},
+                ],
+                "HeaderParameters": [
+                    {"Key": "Content-Type", "Value": "application/x-www-form-urlencoded"},
+                    {"Key": "oauthheader", "Value": "value2"},
+                ],
+                "QueryStringParameters": [{"Key": "oauthquery", "Value": "value3"}],
+            },
+        },
+    },
+]
+
+LAMBDA_BACKEND_CODE = r"""
+import json
+def handler(event, context):
+    print(json.dumps(event))
+    return {
+        "statusCode": 200,
+        "headers": event.get("headers", {}),
+        "body": json.dumps(event.get("body", {})),
+        "queryStringParameters": event.get("queryStringParameters", {})
+    }
+"""
 
-FN_CODE = """
+OAUTH_TOKEN_LAMBDA_CODE = r"""
 import json
 def handler(event, context):
+    print(json.dumps(event))
     return {
         "statusCode": 200,
+        "headers": {"Content-Type": "application/json"},
         "body": json.dumps({
-            "message": "Hello World!"
+            "access_token": "my-oauth-token",
+            "token_type": "bearer",
+            "expires_in": 86400
         })
     }
 """
 
 
 @markers.acceptance_test
-class TestApigatewayLambdaIntegration:
```

Comment:
:thinking: `tests.aws.services.apigateway.test_apigateway_lambda_cfn.TestApiDestinationsCfnComplex` and `tests.aws.services.events.test_api_destinations_cfn.TestApiDestinationsCfn` seem to be the exact same test. I am not sure I understand the purpose of rescoping this test to be an exact copy of another test.

Reply:
Thank you for your inputs @bentsku and @cloutierMat 🙏 

I had a pairing session with @zaingz today to clarify and improve the test design. The goal of the test is to cover the functionality of [Connections for API targets in Amazon EventBridge](https://docs.aws.amazon.com/eventbridge/latest/userguide/eb-target-connection.html). The main suggestions are:
* Replace the validator Lambda with an [API Gateway -> SQS integration](https://docs.aws.amazon.com/prescriptive-guidance/latest/patterns/integrate-amazon-api-gateway-with-amazon-sqs-to-handle-asynchronous-rest-apis.html)
* Replace the oauth Lambda with an API gateway mock integration
* Create three CDK-based tests in EventBridge to avoid mixing CDK-based and boto3-based infrastructure setups (I am unaware of a way to parameterize the infrastructure setup for the three authentication methods).
* Potentially use CDK constructs to avoid duplication between test cases using the authentication type as a parameter. Examples are available in localstack-ext under `localstack-pro-core/tests/aws/scenario/serverlesspresso/constructs`

More details, including a Excalidraw diagram are available in the backlog item "Add aws-validated test for connections behavior"

Reply:
Thanks @joe4dev for the precision on this!

I wonder if it might be easier to use Lambda Function URLs for your setup? API Gateway usually needs a lot of setup, and I'm not sure we need all of this to validate EventBridge Connections.
Your Lambda could always push the message to SQS or save it to S3, however you'd like, but at least you can properly validate the input, SQS integrations are quite annoying to setup to be honest, it's easier to setup a lambda forwarding to SQS (at least it's much easier to shape the output, unlike an SQS integration). 

Regarding mixing Boto3 and CDK, I think this is perfectly fine and much faster than setting up 3 CDK based tests, where the acceptance tests needs to run fast. We're doing it for the DMS service for example, where we have no choice to do so, as long as the test properly cleans up.

Reply:
We have an example for a fixture setting up an external HTTP endpoint using lambda URL with `localstack.testing.pytest.fixtures.create_echo_http_server` for example, it could be adapted/used for you as well I think, or does the infrastructure for EventBridge Connections difficult to setup?

Reply:
@bentsku 

> I wonder if it might be easier to use Lambda Function URLs for your setup? API Gateway usually needs a lot of setup, and I'm not sure we need all of this to validate EventBridge Connections.

My thought was that API Gateway is much faster and its Python-native implementation has fewer dependencies than Lambda, so we are saving CI time in the long run.

> SQS integrations are quite annoying to setup to be honest, it's easier to setup a lambda forwarding to SQS (at least it's much easier to shape the output, unlike an SQS integration).

The output shaping is my main worry here as well. Is it possible to forward the most raw incoming request to SQS or does that only work with Lambda?

> Regarding mixing Boto3 and CDK, I think this is perfectly fine and much faster than setting up 3 CDK based tests, where the acceptance tests needs to run fast. We're doing it for the DMS service for example, where we have no choice to do so, and it's fine as long as the test properly cleans up.

Thanks for sharing @bentsku 👍  Yeah, speed is indeed a good argument. @zaingz feel free to keep the mixed approach as long as we clean up resources entirely.

Reply:
> We have an example for a fixture setting up an external HTTP endpoint using lambda URL with `localstack.testing.pytest.fixtures.create_echo_http_server` for example, it could be adapted and used as an example to create something fitting your use-case as well I think, or is the infrastructure for EventBridge Connections difficult to setup?

The main goal is to have an aws-validated test case for EventBridge Connections. This requires validating the outgoing HTTP request from EventBridge Connections to some receiving HTTP endpoint, which should be able to persist an incoming HTTP somewhere as raw as possible in an asynchronous way (i.e., `put_events` doesn't return the response, hence an echo HTTP server seems insufficient here). What would be your suggestion @bentsku ?

Reply:
If we remove Lambda completely from this stack, probably yes, but it will be at the cost of complexity I would say. 

You could have a class scoped fixture for this specific lambda URL, so it'd be setup only once for all of your tests probably, you should probably do the same with API Gateway as API Gateway takes a lot of HTTP requests to setup a simple API. 

For the input/output shaping for the SQS integration, you will have to use VTL templating, and have to manually declare all wanted incoming headers to be able to pass it to your integration, so it's not the easiest tool to use when you don't know in advance what you want to get. 
As an example, let's say EventBridge will send `X-Amz-Whatever-Header`, if you're not aware of it, it won't be passed to your integration by default, you will have to declare it in your `MethodRequest` so that you can pass it to your `IntegrationRequest`. All in all, I would suggest not to use this kind of integration as it's not really suited for "exploratory" testing where you don't know exactly in advance what you will get as input. 

`AWS_PROXY` integration doesn't have this kind of filtering as it passes the input as-is to your integration. 

All in all, I would suggest to continue using Lambdas to be free of "picking" exactly what you want from the incoming request and just pass everything as-is. And as API Gateway is more complex than Lambda URL, I personally would use Lambda Function URL for your needs, as it fits the initial needs of an externally accessible HTTP endpoint, with much less complexity than a fully fledged REST APIs with an `AWS_PROXY` integration.

Reply:
> The main goal is to have an aws-validated test case for EventBridge Connections. This requires validating the outgoing HTTP request from EventBridge Connections to some receiving HTTP endpoint, which should be able to persist an incoming HTTP somewhere as raw as possible in an asynchronous way

I would setup a lambda Function URL forwarding the incoming event to an SQS queue I think, it keeps the dependencies minimal, only 2 resources created for it (Lambda Function URLs I think uses HTTP APIs under the hood, so it's basically API Gateway already 😄). It's a bit "heavier" if we think about spinning up Lambda Containers, but the flexibility is much greater I would say

Reply:
> I would setup a lambda Function URL forwarding the incoming event to an SQS queue I think, it keeps the dependencies minimal, only 2 resources created for it

That makes a lot of sense
/cc @zaingz 

I wasn't aware that the SQS integration doesn't allow for easy forwarding 💡 , that's definitely not worth the extra complexity here. Sorry for missing out that part @zaingz .

Reply:
> I wasn't aware that the SQS integration doesn't allow for easy forwarding 💡 

No worries, I'll be honest, API Gateway REST APIs are *terrible* to use, the VTL templating, the request input mapping and response output mapping, everything is way too complex. `AWS_PROXY` integrations are alright to use, it's a step in the right direction. 

As a side note, HTTP APIs are much nicer and easier to use, they by default forward all headers to the integration, but they are Pro only here 😬 so not adapted for this test sadly. They have a [`SQS-SendMessage`](https://docs.aws.amazon.com/apigateway/latest/developerguide/http-api-develop-integrations-aws-services-reference.html#SQS-SendMessage) subtype doing what you would think the regular SQS integration should do. Anyway this is out of scope of the discussion, sorry! 😅

Reply:
And the idea of using Lambda Functions URL to have easily accessible external HTTP endpoint came from @cloutierMat, we used it extensively through our API Gateway `HTTP` and `HTTP_PROXY` integration to mimic the behavior of `httpbin.org`, and was initially implemented with #10561, and it really has helped us with different variants of it 😄

Reply:
Let me try to summarize the discussion and recommendations:

* Keep the validator Lambda and use a [Lambda function URL](https://docs.aws.amazon.com/lambda/latest/dg/urls-configuration.html) to expose an HTTP endpoint. The validator Lambda can forward the incoming HTTP request to SQS, allowing easy asynchronous validation (e.g., using the fixture `sqs_collect_messages`). The Lambda approach is less complex than the limited [API Gateway -> SQS integration](https://docs.aws.amazon.com/prescriptive-guidance/latest/patterns/integrate-amazon-api-gateway-with-amazon-sqs-to-handle-asynchronous-rest-apis.html).
* For the oauth Lambda, we can keep and expose it using a function URL or use a lightweight API gateway mock integration returning a hard-coded response.
* Mixing CDK-based base infrastructure setup and executing a boto3-based parametrized test on top is fine 👍  However, we should:
  * Ensure that all resources are properly cleaned up
  * Try to get rid of hardcoded timeouts as much as possible and replace them with waiters

Reply:
Thanks for summarizing! 

I think that if you still want to use API Gateway `MOCK` integration to return hardcoded values, you will go through the trouble of setting up a REST API, so the "validator Lambda" could also be an `AWS_PROXY` integration if you'd prefer, as those are pretty similar.

If you can specify of a path, or the incoming request for the OAuth logic is "recognizable", the validator lambda could also respond to it if we're worried about speed/spawning 2 lambdas. 

(I just realized that the current `oauth` Lambda is already a Function URL 😄) 

Another recommendation would be to leave this API Gateway acceptance test untouched (as we should maybe extend a bit in the future), and create a new acceptance test in `events`, what do you think?